### PR TITLE
frontdoor: refactoring & ensuring ID's are consistent

### LINKED ISSUE #8036: azurerm_frontdoor broken with recent Azure API update
### Terraform (and AzureRM Provider) Version

```
Terraform v0.12.29
+ provider.azurerm v2.21.0
```

### Affected Resource(s)

* `azurerm_frontdoor`

### Terraform Configuration Files

```hcl
resource "azurerm_frontdoor" "frontdoor" {
    name = "bink-frontdoor"
    resource_group_name = azurerm_resource_group.rg.name
    enforce_backend_pools_certificate_name_check = true

    tags = var.tags

< snipped >

    routing_rule {
        name = "api-prod-<snipped>"
        accepted_protocols = ["Https"]
        patterns_to_match = ["/*"]
        frontend_endpoints = ["default", "api-gb-bink-com", "api-bink-com"]
        forwarding_configuration {
            forwarding_protocol = "HttpsOnly"
            backend_pool_name = "api-prod-<snipped>"
            cache_enabled = false
        }
    }

< snipped >

}
```

### Debug Output

Terraform Plan Output:

```
      ~ routing_rule {
            accepted_protocols = [
                "Https",
            ]
            enabled            = true
          ~ frontend_endpoints = [
              - "",
              - "",
              - "",
              + "default",
              + "api-gb-bink-com",
              + "api-bink-com",
            ]
            id                 = "/subscriptions/0add5c8e-<snipped>/resourcegroups/frontdoor/providers/Microsoft.Network/Frontdoors/bink-frontdoor/RoutingRules/api-prod-<snipped>"
            name               = "api-prod-<snipped>"
            patterns_to_match  = [
                "/*",
            ]

          ~ forwarding_configuration {
              + backend_pool_name                     = "api-prod-<snipped>"
                cache_enabled                         = false
                cache_query_parameter_strip_directive = "StripAll"
                cache_use_dynamic_compression         = false
                forwarding_protocol                   = "HttpsOnly"
            }
        }
```

Azure CLI Output:

```json
  {
    "acceptedProtocols": [
      "Https"
    ],
    "enabledState": "Enabled",
    "frontendEndpoints": [
      {
        "id": "/subscriptions/0add5c8e-<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/frontendendpoints/default",
        "resourceGroup": "frontdoor"
      },
      {
        "id": "/subscriptions/0add5c8e-<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/frontendendpoints/api-gb-bink-com",
        "resourceGroup": "frontdoor"
      },
      {
        "id": "/subscriptions/0add5c8e-<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/frontendendpoints/api-bink-com",
        "resourceGroup": "frontdoor"
      }
    ],
    "id": "/subscriptions/0add5c8e-<snipped>/resourcegroups/frontdoor/providers/Microsoft.Network/Frontdoors/bink-frontdoor/RoutingRules/api-prod-<snipped>",
    "name": "api-prod-<snipped>",
    "patternsToMatch": [
      "/*"
    ],
    "resourceGroup": "frontdoor",
    "resourceState": "Enabled",
    "routeConfiguration": {
      "backendPool": {
        "id": "/subscriptions/0add5c8e-<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/backendpools/api-prod-<snipped>",
        "resourceGroup": "frontdoor"
      },
      "cacheConfiguration": null,
      "customForwardingPath": null,
      "forwardingProtocol": "HttpsOnly",
      "odatatype": "#Microsoft.Azure.FrontDoor.Models.FrontdoorForwardingConfiguration"
    },
    "rulesEngine": null,
    "type": "Microsoft.Network/Frontdoors/RoutingRules",
    "webApplicationFirewallPolicyLink": null
  },
```

### Expected Behavior

Terraform identifies that no changes are required

### Actual Behavior

Azure API is responding with:

`"id": "/subscriptions/0add5c8e-<snipped>/resourcegroups/frontdoor/providers/microsoft.network/frontdoors/bink-frontdoor/frontendendpoints/api-gb-bink-com"`

instead of

`"id": "/subscriptions/0add5c8e-<snipped>/resourcegroups/frontdoor/providers/Microsoft.Network/Frontdoors/bink-frontdoor/FrontendEndpoints/api-gb-bink-com"`

Which terraform is identifying as different due to the lack of capitalisation.

### Steps to Reproduce

Attempt to spin up a Front Door and Apply/Plan a second time

### Important Factoids

Have raised a P1 with the Azure Support team to investigate as this is clearly a breaking API change.

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR introduces quite a large refactoring to the Frontdoor resources - namely this ensures that all fields are always set - and ensures that ID's are consistent, so that users can rely on these.

Supersedes #8046
Fixes #8036

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2020-08-17 at 18 48 56

File: azurerm/internal/services/frontdoor/validate/name_test.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		Input       string
+		ExpectError bool
+	}{
+		{
+			Name:        "Empty String",
+			Input:       "",
+			ExpectError: true,
+		},
+		{
+			Name:        "Starst with Numeric",
```
Comment (jackofallops) [inline azurerm/internal/services/frontdoor/validate/name_test.go:19]:
```suggestion
			Name:        "Starts with Numeric",
```

File: azurerm/internal/services/frontdoor/frontdoor_firewall_policy_resource.go [hunk 2]
```diff
@@ -12,23 +12,29 @@ import (
...
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor/migration"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor/parse"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor/validate"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tags"
+	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
 )
 
+// TODO: a state migration to patch the ID's
```
Comment (jackofallops) [inline azurerm/internal/services/frontdoor/frontdoor_firewall_policy_resource.go:25]: This looks to have been done?

File: azurerm/internal/services/frontdoor/parse/frontdoor.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		return nil, err
+	}
+
+	return frontDoorId, nil
+}
+
+func FrontDoorIDForImport(input string) (*FrontDoorId, error) {
+	id, err := azure.ParseAzureResourceID(input)
+	if err != nil {
+		return nil, fmt.Errorf("[ERROR] Unable to parse FrontDoor ID %q: %+v", input, err)
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/internal/services/frontdoor/parse/frontdoor.go:38]: will make this "parsing"

File: azurerm/internal/services/frontdoor/parse/frontdoor.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	return FrontDoorId{
+		Name:          name,
+		ResourceGroup: resourceGroup,
+	}
+}
+
+func FrontDoorID(input string) (*FrontDoorId, error) {
+	frontDoorId, id, err := parseFrontDoorChildResourceId(input)
+	if err != nil {
+		return nil, fmt.Errorf("[ERROR] Unable to parse FrontDoor ID %q: %+v", input, err)
```
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/internal/services/frontdoor/parse/frontdoor.go:25]: (as below) will make this "parsing"

**Pull request merged**